File: system-test/src/main/java/com/jcupac/bookshop/testkit/dsl/core/usecase/usecases/GoToBookShop.java
package com.jcupac.bookshop.testkit.dsl.core.usecase.usecases;

import com.jcupac.bookshop.testkit.driver.port.BookShopDriver;
import com.jcupac.bookshop.testkit.driver.port.dtos.GoToBookShopRequest;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.base.BaseBookShopUseCase;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseResult;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseContext;
import com.jcupac.bookshop.testkit.dsl.core.shared.VoidVerification;

public class GoToBookShop extends BaseBookShopUseCase<Void, VoidVerification> {
    public GoToBookShop(BookShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var request = GoToBookShopRequest.builder().build();
        var result = driver.goToBookShop(request).mapVoid();
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}

File: system-test/src/testkit/dsl/core/usecase/usecases/GoToBookShop.ts
import type { BookShopDriver } from '../../../../driver/port/book-shop-driver.js';
import type { GoToBookShopResponse } from '../../../../driver/port/dtos/GoToBookShopResponse.js';
import { UseCaseResult } from '../../shared/use-case-result.js';
import { VoidVerification } from '../../shared/void-verification.js';
import type { UseCaseContext } from '../../shared/use-case-context.js';
import { BaseBookShopUseCase } from './base/BaseBookShopUseCase.js';

export class GoToBookShop extends BaseBookShopUseCase<GoToBookShopResponse, VoidVerification> {
  constructor(driver: BookShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  async execute(): Promise<UseCaseResult<GoToBookShopResponse, VoidVerification>> {
    const result = await this.driver.goToBookShop({});

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
